datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -70,29 +74,33 @@ use hashbrown::raw::RawTable;
 ///
 /// [Compact]: datafusion_row::layout::RowType::Compact
 /// [WordAligned]: datafusion_row::layout::RowType::WordAligned
-pub(crate) struct GroupedHashAggregateStreamV2 {
+pub(crate) struct GroupedHashAggregateStream {
     stream: BoxStream<'static, ArrowResult<RecordBatch>>,
     schema: SchemaRef,
 }
 
-/// Actual implementation of [`GroupedHashAggregateStreamV2`].
+/// Actual implementation of [`GroupedHashAggregateStream`].
 ///
 /// This is wrapped into yet another struct because we need to interact with the async memory management subsystem
 /// during poll. To have as little code "weirdness" as possible, we chose to just use [`BoxStream`] together with
-/// [`futures::stream::unfold`]. The latter requires a state object, which is [`GroupedHashAggregateStreamV2Inner`].
-struct GroupedHashAggregateStreamV2Inner {
+/// [`futures::stream::unfold`]. The latter requires a state object, which is [`GroupedHashAggregateStreamInner`].
+struct GroupedHashAggregateStreamInner {
     schema: SchemaRef,
     input: SendableRecordBatchStream,
     mode: AggregateMode,
-    aggr_state: AggregationState,
-    aggregate_expressions: Vec<Vec<Arc<dyn PhysicalExpr>>>,
+    normal_aggr_expr: Vec<Arc<dyn AggregateExpr>>,
+    row_aggr_state: RowAggregationState,
+    /// Aggregate expressions not supporting row accumulation
+    normal_aggregate_expressions: Vec<Vec<Arc<dyn PhysicalExpr>>>,
```

Comment:
So we have one operator now but still two accumulator implementations? I think think this counts as "closing the ticket" though. We still have massive code duplication in accumulator implementations. Furthermore the implementation of this specific stream gets more complicated, but I agree that this would be one possible path forward.

Reply:
Yes, one operator but multiple accumulators. Having a single accumulator implementation without performance impact is a tough goal and there seems to be some open questions around it. For example, some accumulation tasks involve non-fixed states, how do we implement them in the row-accumulator style?

Until we see a clear path forward in that regard, I think this is a good first step towards full unification (if such a thing is possible).

Reply:
See https://github.com/apache/arrow-datafusion/issues/2723#issuecomment-1324876060

Reply:
I.e. I'm personally convinced that the row accumulator style is a dead end.

Reply:
I will try and help organize the tickets some more tomorrow morning. Thanks for this discussion